### PR TITLE
added support to sign linux artifacts

### LINKED ISSUE #1351: [META] Automate artifact signing with OpensearchSignerClient for linux
### Is your feature request related to a problem? Please describe

Signing currently is a long painstaking process. It needs to be automated to save time when releasing artifacts. We need a tool to be able to sign all artifacts - 
- opensearch
- opensearch-dashboards

### Describe the solution you'd like

Use the existing `Signer.py` class to sign the artifacts and generate ".sig" file along with ".asc" files.

### Tasks
- [x] #1382
- [x] #1383
- [x] #1385

### Acceptance Criteria
- [ ] User is able to provide the artifact directory and tool is able to sign all the artifacts in the directory
- [ ] User is able to pass the artifact path and the tool is able to sign the artifact

### Next Steps
Next steps would include to extend this process over for mac and windows

### PR DESCRIPTION
**Pull request by abhinavGupta16**
Signed-off-by: Abhinav Gupta <abhng@amazon.com>

### Description
Automate linux signing to sign any artifact or all artifacts in a directory
 
### Issues Resolved
#1351
#1382

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/sign_workflow/sign_existing_artifacts.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    def __init__(self, artifact_path, signature_type):
+        self.artifact_path = artifact_path
+        self.signature_type = signature_type
+
+    def __setattr__(self, name, value):
+        if name == 'artifact_path' and not isinstance(value, Path):
+            raise TypeError('SignExistingArtifacts.artifact_path must of type Path')
+        super().__setattr__(name, value)
+
+    def sign_using_manifest(self):
```
Comment (gaiksaya) [inline src/sign_workflow/sign_existing_artifacts.py:26]: did you mean `sign_existing_artifacts` here? using manifest already exists in https://github.com/opensearch-project/opensearch-build/pull/1352/files#diff-4bf5b6c3689571f89b11c463c1a6a4ff6bcdceb0a5e0ed1726d23886cb39ea18R24

Comment (abhinavGupta16) [reply src/sign_workflow/sign_existing_artifacts.py:26]: No, It is `SignExistingArtifacts` as in class name.

File: src/sign_workflow/sign_existing_artifacts.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+class SignExistingArtifacts:
+
+    def __init__(self, artifact_path, signature_type):
+        self.artifact_path = artifact_path
+        self.signature_type = signature_type
+
+    def __setattr__(self, name, value):
+        if name == 'artifact_path' and not isinstance(value, Path):
+            raise TypeError('SignExistingArtifacts.artifact_path must of type Path')
+        super().__setattr__(name, value)
```
Comment (gaiksaya) [inline src/sign_workflow/sign_existing_artifacts.py:24]: Why not add `artifact_path : Path` instead of this?

File: src/sign_workflow/sign_existing_artifacts.py (hunk 1 above)
Comment (gaiksaya) [reply src/sign_workflow/sign_existing_artifacts.py:26]: why both methods are named same?

Comment (abhinavGupta16) [reply src/sign_workflow/sign_existing_artifacts.py]: fixed

File: src/sign_workflow/sign_existing_artifacts.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    def sign_single_artifacts(self):
+        signer = Signer()
+        artifacts = [self.artifact_path.name]
+        basename = self.artifact_path.parent
+        signer.sign_artifacts(artifacts, basename, self.signature_type)
+
+    def sign_all_artifacts_in_dir(self):
+        signer = Signer()
+        files_map = {}
```
[the remark this answers is not in the document]
Comment (abhinavGupta16) [reply src/sign_workflow/sign_existing_artifacts.py:37]: No, this cannot be combined since the Signer class takes one basepath and corresponding list of artifacts to sign as inputs.

File: src/run_sign.py [hunk 4]
```diff
@@ -8,19 +8,22 @@
...
+ACCEPTED_SIGNATURE_FILE_TYPES = [".sig", ".asc"]
+
 
 def main():
     parser = argparse.ArgumentParser(description="Sign artifacts")
-    parser.add_argument("manifest", type=argparse.FileType("r"), help="Path to local manifest file.")
+    parser.add_argument("target", type=Path, help="Path to local manifest file or artifact directory")
     parser.add_argument("--component", nargs="?", help="Component name")
     parser.add_argument("--type", nargs="?", help="Artifact type")
+    parser.add_argument("--sigType", nargs="?", help="Type of Signature file", default=".asc")
```
Comment (dblock) [inline src/run_sign.py:26]: Arguments shouldn’t be camelcase, —sigtype or —signature-type.
Lowercase signature in the description.

File: src/run_sign.py [hunk 5]
```diff
@@ -32,22 +35,20 @@ def main():
     )
     args = parser.parse_args()
 
-    console.configure(level=args.logging_level)
-
-    manifest = BuildManifest.from_file(args.manifest)
-    basepath = os.path.dirname(os.path.abspath(args.manifest.name))
-    signer = Signer()
+    if args.sigType not in ACCEPTED_SIGNATURE_FILE_TYPES:
```
Comment (dblock) [inline src/run_sign.py:38]: There’s a way to add a list of allowed values to args and remove this.

File: src/run_sign.py [hunk 6]
```diff
@@ -32,22 +35,20 @@ def main():
...
-        for artifact_type in component.artifacts:
-            if args.type and args.type != artifact_type:
-                continue
+    console.configure(level=args.logging_level)
 
-            signer.sign_artifacts(component.artifacts[artifact_type], basepath)
+    file_extension = args.target.suffix
 
-    logging.info("Done.")
+    if file_extension == ".yml":
```
Comment (dblock) [inline src/run_sign.py:45]: Add a class called Signers that has a method called from_file that returns either SignWithManifest or SignArtifacts. Both should inherit from the same base class and have a .sign method that you call once. The duplicate logic of creating a Signer should also be refactored - I think you want to get a list of files to sign in different ways but then sign each item in the list in the base class.

**Review by dblock — CHANGES_REQUESTED** [review]

File: src/sign_workflow/sign_existing_artifacts.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (peternied) [reply src/sign_workflow/sign_existing_artifacts.py:37]: I am pretty you can but this is nit

File: src/run_sign.py [hunk 7]
```diff
@@ -8,19 +8,22 @@
...
 import sys
+from pathlib import Path
 
-from manifests.build_manifest import BuildManifest
-from sign_workflow.signer import Signer
+from sign_workflow.sign_existing_artifacts import SignExistingArtifacts
+from sign_workflow.sign_with_manifest import SignWithManifest
 from system import console
 
+ACCEPTED_SIGNATURE_FILE_TYPES = [".sig", ".asc"]
```
[the remark this answers is not in the document]
Comment (abhinavGupta16) [reply src/run_sign.py:18]: The existing workflow used for maven currently is the same as it is here, which only renames the binary file to be `.asc` and does not create new `.asc` files. They are therefore essentially `.sig` binary files.
Here, I have added support so that the existing maven workflow would remain the same and we can also create `.sig` files when required (eg: opensearch)

File: src/run_sign.py [hunk 8]
```diff
@@ -8,19 +8,22 @@
...
+
 
 def main():
     parser = argparse.ArgumentParser(description="Sign artifacts")
-    parser.add_argument("manifest", type=argparse.FileType("r"), help="Path to local manifest file.")
+    parser.add_argument("target", type=Path, help="Path to local manifest file or artifact directory")
```
[the remark this answers is not in the document]
Comment (dblock) [reply src/run_sign.py:23]: Just consistency.

File: src/sign_workflow/sign_existing_artifacts.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (abhinavGupta16) [reply src/sign_workflow/sign_existing_artifacts.py]: nvm... got it!

File: src/run_sign.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (peternied) [reply src/run_sign.py:18]: The issue is this change does not make properly produced `.asc` files and hides the fact they are not generated correctly.  We should either apply the workaround or not support `.asc` until it works properly

[the remark this answers is not in the document]
Comment (peternied) [reply src/run_sign.py:18]: Issue in our codebase #368

File: src/sign_workflow/sign_existing_artifacts.py [hunk 9]
```diff
@@ -34,13 +34,8 @@ def sign_single_artifacts(self):
...
+            artifacts = []
             for file in files:
-                if subdir in files_map:
-                    files_map[subdir].append(file)
-                else:
-                    files_map[subdir] = [file]
-
-        for basename, artifact_list in files_map.items():
-            signer.sign_artifacts(artifact_list, basename, self.signature_type)
+                artifacts.append(file)
```
Comment (dblock) [inline src/sign_workflow/sign_existing_artifacts.py:40]: Isn't this just `files`, so `signer.sign_artifacts(files, subdir, self.signature_type)`

File: src/run_sign.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (abhinavGupta16) [reply src/run_sign.py:18]: We can work on #368 in a different PR and maybe use `gpg enarmour` as a workaround?

[the remark this answers is not in the document]
Comment (peternied) [reply src/run_sign.py:17]: [Blocking] To keep this PR mergable remove the ASC option, we can follow up on that open issue to add it back in the future.

You don't need to remove the plumbing, just the option from the array

File: tests/test_run_sign.py [hunk 10]
```diff
@@ -32,7 +32,7 @@ def test_usage(self, *mocks):
 
     @patch("os.getcwd", return_value="curdir")
     @patch("argparse._sys.argv", ["run_sign.py", BUILD_MANIFEST])
-    @patch("run_sign.Signer", return_value=MagicMock())
+    @patch("sign_workflow.sign_artifacts.SignArtifacts", return_value=MagicMock())
```
Comment (peternied) [inline tests/test_run_sign.py:35]: How did you figure out the trick to get this working?

File: vars/signArtifacts.groovy [hunk 11]
```diff
@@ -0,0 +1,25 @@
+void call(Map args = [:]) {
```
Comment (peternied) [inline vars/signArtifacts.groovy:1]: Lets moving this change into its own PR so it doesn't impact the python workflow that is ready to merge

File: vars/signArtifacts.groovy [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+    environment {
+        // These ENV variables are required by https://github.com/opensearch-project/opensearch-signer-client
+        // This client is invoked internally by the sign script.
+        ROLE = "${SIGNER_CLIENT_ROLE}"
+        EXTERNAL_ID = "${SIGNER_CLIENT_EXTERNAL_ID}"
+        UNSIGNED_BUCKET = "${SIGNER_CLIENT_UNSIGNED_BUCKET}"
+        SIGNED_BUCKET = "${SIGNER_CLIENT_SIGNED_BUCKET}"
+    }
+
+    git url: 'https://github.com/opensearch-project/opensearch-build.git', branch: 'main'
```
Comment (peternied) [inline vars/signArtifacts.groovy:12]: Will this have an issue if run multiple times without cleaning the workspace?

File: vars/signArtifacts.groovy [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+
+    sh('curl https://artifacts.opensearch.org/publickeys/opensearch.pgp -o $WORKSPACE/opensearch.pgp')
+    sh('gpg --import $WORKSPACE/opensearch.pgp')
+
+    def signatureType = ".sig"
```
Comment (peternied) [inline vars/signArtifacts.groovy:17]: The value should come from args.signatureType not hard coded (same for other parameters)

See https://github.com/opensearch-project/opensearch-build/blob/50552294f6830d7412d66831798fe15acf89e43d/vars/buildManifest.groovy for an example.

File: vars/signArtifacts.groovy [hunk 14]
```diff
@@ -0,0 +1,25 @@
+void call(Map args = [:]) {
+
```
Comment (peternied) [inline vars/signArtifacts.groovy:2]: [For the end to end scenario, not this PR]

In order to call this function from the promotion script it needs to be in the form of a closure - that part can come afterward when the sha512 changes are made.

It would could make sense to keep this file as is built and create another function that returns the closure version.

**Review by peternied — CHANGES_REQUESTED** [review]

File: src/run_sign.py [hunk 15]
```diff
@@ -8,19 +8,21 @@
...
-import os
 import sys
+from pathlib import Path
 
-from manifests.build_manifest import BuildManifest
-from sign_workflow.signer import Signer
+from sign_workflow.sign_artifacts import SignArtifacts
 from system import console
 
+ACCEPTED_SIGNATURE_FILE_TYPES = [".sig", ".asc"]
```
Comment (dblock) [inline src/run_sign.py:17]: Isn't this already `Signer.ACCEPTED_FILE_TYPES`? Use that.

File: src/run_sign.py [hunk 16]
```diff
@@ -34,20 +36,10 @@ def main():
...
-
-    for component in manifest.components.select(focus=args.component):
-        logging.info(f"Signing {component.name}")
-
-        for artifact_type in component.artifacts:
-            if args.type and args.type != artifact_type:
-                continue
-
-            signer.sign_artifacts(component.artifacts[artifact_type], basepath)
+    sign = SignArtifacts.from_path(path=args.target, component=args.component,
```
Comment (dblock) [inline src/run_sign.py:39]: nit: put each arg on its own line, it's easier to diff when they change

File: src/sign_workflow/sign_artifacts.py [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+
+    @abstractmethod
+    def __sign__(self):
+        pass
+
+    def sign(self):
+        self.__sign__()
+        logging.info("Done.")
+
+    def __sign_artifacts__(self, artifacts, basepath, signature_type):
```
Comment (dblock) [inline src/sign_workflow/sign_artifacts.py:34]: The `signature_type` argument here is redundant because it's already passed into the constructor, so you can do `self.signer.sign_artifacts(artifacts, basepath, self.signature_type)`. Remove it.

File: src/sign_workflow/sign_artifacts.py [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+        else:
+            return SignArtifactsExistingArtifactFile
+
+    @classmethod
+    def from_path(self, path: Path, component, artifact_type, signature_type):
+        klass = self.__signer_class__(path)
+        return klass(path, component, artifact_type, signature_type)
+
+
+class SignWithManifest(SignArtifacts):
```
Comment (dblock) [inline src/sign_workflow/sign_artifacts.py:52]: rename to `SignWithBuildManifest` to be clear what kind of manifest we expect

File: src/sign_workflow/sign_artifacts.py [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+                    continue
+
+                super().__sign_artifacts__(component.artifacts[component_artifact_type], basepath, self.signature_type)
+
+
+class SignArtifactsExistingArtifactFile(SignArtifacts):
+
+    def __sign__(self):
+        artifacts = [self.target.name]
+        basename = self.target.parent
```
Comment (dblock) [inline src/sign_workflow/sign_artifacts.py:71]: These variables aren't reused, just pass the arguments as is below.

Add `__sign_artifact__` that takes one artifact.

File: src/sign_workflow/signer.py [hunk 20]
```diff
@@ -27,32 +26,27 @@ def __init__(self):
         self.git_repo.execute("./bootstrap")
         self.git_repo.execute("rm config.cfg")
 
-    def sign_artifacts(self, artifacts, basepath):
+    def sign_artifacts(self, artifacts, basepath, signature_type):
```
Comment (dblock) [inline src/sign_workflow/signer.py:29]: There's a caller that passes 1 file and another that passes a collection. Split this function in 2, one that receives a single file and another that receives a collection.

File: src/sign_workflow/sign_artifacts.py (hunk 17 above)
Comment (dblock) [inline src/sign_workflow/sign_artifacts.py:34]: Add a single `__sign_artifact__` to avoid having to make an array in the caller.

File: tests/tests_sign_workflow/test_sign_artifacts.py [hunk 21]
```diff
@@ -0,0 +1,40 @@
+import unittest
+from pathlib import Path
+
+from sign_workflow.sign_artifacts import SignArtifacts, SignArtifactsExistingArtifactFile, SignExistingArtifactsDir, SignWithManifest
+
+
+class TestSignArtifacts(unittest.TestCase):
+
+    def test_signer_class_method(self):
```
Comment (dblock) [inline tests/tests_sign_workflow/test_sign_artifacts.py:9]: Remove `_class_method` from the name of the function, use the actual function name.

File: tests/tests_sign_workflow/test_sign_artifacts.py [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+        path = Path(r"/dummy/path/manifest.yml")
+        klass = SignArtifacts.__signer_class__(path)
+
+        self.assertEqual(type(SignWithManifest), type(klass))
+
+        path = Path(r"/dummy/path/")
```
Comment (dblock) [inline tests/tests_sign_workflow/test_sign_artifacts.py:15]: Split up in separate tests.

File: tests/tests_sign_workflow/test_sign_artifacts.py [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+
+        self.assertEqual(type(SignArtifactsExistingArtifactFile), type(klass))
+
+    def test_from_path_method(self):
+        path = Path(r"/dummy/path/manifest.yml")
+        component = 'maven'
+        artifact_type = 'dummy'
+        sigtype = '.asc'
+
+        klass = SignArtifacts.from_path(path, component, artifact_type, sigtype)
```
Comment (dblock) [inline tests/tests_sign_workflow/test_sign_artifacts.py:31]: Same as above, split up into individual tests for each type.

**Review by dblock — CHANGES_REQUESTED** [review]
Close!

File: tests/test_run_sign.py (hunk 10 above)
Comment (abhinavGupta16) [reply tests/test_run_sign.py:35]: It's not working yet.

File: src/run_sign.py (hunk 15 above)
Comment (abhinavGupta16) [reply src/run_sign.py:17]: No, this is accepted Signature types, different from accepted file types

File: src/sign_workflow/sign_artifacts.py (hunk 19 above)
Comment (abhinavGupta16) [reply src/sign_workflow/sign_artifacts.py:71]: Variables just make it clear as to what is being passed to the function. I think we can leave it like this

File: tests/tests_sign_workflow/test_sign_artifacts.py (hunk 21 above)
Comment (abhinavGupta16) [reply tests/tests_sign_workflow/test_sign_artifacts.py:9]: The name of the method is `__signer_class__`

File: tests/tests_sign_workflow/test_sign_artifacts.py (hunk 22 above)
Comment (abhinavGupta16) [reply tests/tests_sign_workflow/test_sign_artifacts.py:15]: I think it should be one method, since it just verifies multiple if conditions with no workflow. It would just increase verbosity and complexity. This is clear too. Thoughts?

File: tests/tests_sign_workflow/test_sign_artifacts.py (hunk 23 above)
Comment (abhinavGupta16) [reply tests/tests_sign_workflow/test_sign_artifacts.py:31]: Same reason as before.

File: tests/tests_sign_workflow/test_sign_artifacts.py (hunk 21 above)
Comment (dblock) [reply tests/tests_sign_workflow/test_sign_artifacts.py:9]: It would then be called `test_signer_class`.

You should use `assertIs` to avoid to do `type` twice.

Using variables that aren't actually reused is an anti-pattern.

Maybe this reads cleaner?

```python
    def test_signer_class(self):
        self.assertIs(SignArtifacts.__signer_class__(
            Path(r"/dummy/path/manifest.yml")), 
            SignWithBuildManifest)
            
        self.assertIs(SignArtifacts.__signer_class__(
            Path(os.path.dirname(__file__))), 
            SignExistingArtifactsDir)

        self.assertIs(SignArtifacts.__signer_class__(
            Path(r"/dummy/path/artifact.tar.gz")), 
            SignArtifactsExistingArtifactFile)
```

File: tests/tests_sign_workflow/test_sign_artifacts.py (hunk 22 above)
Comment (dblock) [reply tests/tests_sign_workflow/test_sign_artifacts.py:15]: It's not a must have, but my reasoning is that these tests can fail independently, and methods are cheap.

File: tests/tests_sign_workflow/test_sign_artifacts.py [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+        klass = SignArtifacts.__signer_class__(path)
+
+        self.assertEqual(type(SignExistingArtifactsDir), type(klass))
+
+        path = Path(r"/dummy/path/artifact.tar.gz")
+        klass = SignArtifacts.__signer_class__(path)
+
+        self.assertEqual(type(SignArtifactsExistingArtifactFile), type(klass))
+
+    def test_from_path_method(self):
```
Comment (dblock) [inline tests/tests_sign_workflow/test_sign_artifacts.py:25]: This ends up instantiating an instance of `Signer` and asks for git credentials. You have three options:

- patch `Signer` so it becomes a mock, this is preferred 
- make it a noop with `@patch("sign_workflow.signer.GitRepository")`, this is not recommended because it implies knowledge of what `Signer` is doing, one level too deep
- remove that instantiating from the constructor of `SignArtifacts`, which is not recommended because it will be executing `git checkout` every time we sign anything

**Review by peternied — APPROVED** [review]
Thanks for addressing my comments, great work enhancing this flow.

**Comment by abhinavGupta16** [comment]
Thanks @dblock for helping out with the test cases

File: tests/test_run_sign.py (hunk 10 above)
Comment (abhinavGupta16) [reply tests/test_run_sign.py]: It's fixed now

File: src/sign_workflow/signer.py [hunk 25]
```diff
@@ -19,40 +18,40 @@
...
 
     def is_valid_file_type(self, file_name):
         return any(
-            x
-            in [
-                pathlib.Path(file_name).suffix,
-                "".join(pathlib.Path(file_name).suffixes),
-            ]
-            for x in Signer.ACCEPTED_FILE_TYPES
+            file_name.endswith(x) for x in Signer.ACCEPTED_FILE_TYPES
```
Comment (dblock) [inline src/sign_workflow/signer.py:45]: Use `os.path.splitext` to get the extension from the file name, then `in` to check ether it’s one of the accepted types.

**Review by dblock — APPROVED** [review]
Works for me.

I would also remove all the passing around of `Signer()` which is always the same class, and create an instance of it at the lowest level.

File: src/sign_workflow/signer.py (hunk 25 above)
Comment (abhinavGupta16) [reply src/sign_workflow/signer.py:45]: That does not work since we have artifacts with names like `opensearch-1.0.0.tar.gz`. That is why I changed it to `endswith`

Comment (dblock) [reply src/sign_workflow/signer.py:45]: We could just change .tar.gz into .gz. Technically the extension of that file is not .tar.gz, it’s .gz.
